File: src/main/java/org/mitallast/queue/transport/http/HttpServerHandler.java
package org.mitallast.queue.transport.http;

import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import io.netty.handler.codec.http.DefaultFullHttpResponse;
import io.netty.handler.codec.http.FullHttpRequest;
import io.netty.handler.codec.http.FullHttpResponse;
import io.netty.handler.codec.http.HttpResponseStatus;
import io.netty.util.CharsetUtil;
import org.mitallast.queue.rest.RestController;
import org.mitallast.queue.transport.TransportRequest;
import org.mitallast.queue.transport.TransportSession;

import static io.netty.handler.codec.http.HttpHeaders.Names.CONTENT_TYPE;
import static io.netty.handler.codec.http.HttpResponseStatus.BAD_REQUEST;
import static io.netty.handler.codec.http.HttpResponseStatus.INTERNAL_SERVER_ERROR;
import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;

@ChannelHandler.Sharable
public class HttpServerHandler extends SimpleChannelInboundHandler<FullHttpRequest> {

    private final RestController restController;

    public HttpServerHandler(RestController restController) {
        this.restController = restController;
    }


    private static void sendError(ChannelHandlerContext ctx, HttpResponseStatus status) {
        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, status, Unpooled.copiedBuffer("Failure: " + status.toString() + "\r\n", CharsetUtil.UTF_8));
        response.headers().set(CONTENT_TYPE, "text/plain; charset=UTF-8");
        // Close the connection as soon as the error message is sent.
        ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
        System.out.println(status.toString());
    }

    @Override
    protected void messageReceived(ChannelHandlerContext ctx, FullHttpRequest httpRequest) throws Exception {
        if (!httpRequest.getDecoderResult().isSuccess()) {
            sendError(ctx, BAD_REQUEST);
            return;
        }

        final TransportRequest request = new TransportRequest(httpRequest);
        restController.dispatchRequest(request, new TransportSession(ctx, httpRequest));
    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
        ctx.flush();
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        cause.printStackTrace();
        if (ctx.channel().isActive()) {
            sendError(ctx, INTERNAL_SERVER_ERROR);
        }
        ctx.close();
    }
}

File: src/main/java/org/mitallast/queue/rest/netty/codec/server/HttpServerHandler.kt
package org.mitallast.queue.rest.netty.codec.server

import com.google.inject.Inject
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelHandler
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import io.netty.util.CharsetUtil
import org.apache.logging.log4j.LogManager
import org.mitallast.queue.rest.netty.codec.*
import java.nio.charset.StandardCharsets

@Suppress("OverridingDeprecatedMember")
@ChannelHandler.Sharable
class HttpServerHandler @Inject constructor() : SimpleChannelInboundHandler<HttpRequest>(false) {
    private val logger = LogManager.getLogger(HttpServerHandler::class.java)

    override fun isSharable(): Boolean {
        return true
    }

    override fun channelRegistered(ctx: ChannelHandlerContext) {
        logger.info("channel registered: {}", ctx.channel())
        super.channelActive(ctx)
    }

    override fun channelInactive(ctx: ChannelHandlerContext) {
        logger.info("channel inactive: {}", ctx.channel())
        super.channelInactive(ctx)
    }

    override fun channelRead0(ctx: ChannelHandlerContext, httpRequest: HttpRequest) {
        val content = ctx.alloc().buffer()
        content.writeCharSequence("hello world", StandardCharsets.UTF_8)

        val response = HttpResponse(
            httpRequest.version(),
            HttpResponseStatus.OK,
            HttpHeaders.newInstance(),
            content
        )
        response.headers().put(HttpHeaderName.CONTENT_LENGTH, AsciiString.of(Integer.toString(content.readableBytes())))
        if (httpRequest.version() == HttpVersion.HTTP_1_0) {
            response.headers().put(HttpHeaderName.CONNECTION, HttpHeaderValue.CLOSE.ascii())
            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE)
        } else {
            ctx.write(response, ctx.voidPromise())
        }
        httpRequest.release()
    }

    override fun channelReadComplete(ctx: ChannelHandlerContext) {
        ctx.flush()
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        logger.error("unexpected exception", cause)
        if (ctx.channel().isActive) {
            sendError(ctx, HttpResponseStatus.INTERNAL_SERVER_ERROR)
        }
        ctx.close()
    }

    private fun sendError(ctx: ChannelHandlerContext, status: HttpResponseStatus) {
        val response = HttpResponse(
            HttpVersion.HTTP_1_0,
            status,
            HttpHeaders.newInstance(),
            Unpooled.copiedBuffer("Failure: " + status.toString() + "\r\n", CharsetUtil.UTF_8))
        response.headers().set(HttpHeaderName.CONTENT_TYPE, AsciiString.of("text/plain; charset=UTF-8"))
        response.headers().set(HttpHeaderName.CONTENT_LENGTH, AsciiString.of(Integer.toString(response.content().readableBytes())))
        ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE)
    }
}
